graspy/inference/latent_distribution_test.py

```diff
@@ -14,121 +14,178 @@
 
-        if bandwidth is not None and not isinstance(bandwidth, float):
-            msg = "bandwidth must an int, not {}".format(type(bandwidth))
+        if not isinstance(num_workers, int):
+            msg = "num_workers must be an int, not {}".format(type(num_workers))
             raise TypeError(msg)
+        elif num_workers <= 0:
+            msg = "{} is invalid number of workers, must be greater than 0"
+            raise ValueError(msg.format(num_workers))
+        elif num_workers > 1:
+            raise NotImplementedError()  # TODO env error parallelizing
 
         super().__init__(embedding="ase", n_components=n_components)
+
+        if callable(metric):
+            self.test = KSample(test, compute_distance=metric)
+        else:
+            if metric == "gaussian":
+                self.test = KSample(test, compute_distance=self._medial_gaussian_kernel)
+            else:
+
+                # TODO workers is a parameter required by hyppo.
+                # will be fixed in later releases.
+                def dist_func(X, Y=None, metric=metric, workers=None):
+                    return pairwise_distances(X, Y, metric=metric)
+
+                self.test = KSample(test, compute_distance=dist_func)
+
         self.n_bootstraps = n_bootstraps
-        self.bandwidth = bandwidth
-
-    def _gaussian_covariance(self, X, Y):
-        diffs = np.expand_dims(X, 1) - np.expand_dims(Y, 0)
-        if self.bandwidth is None:
-            self.bandwidth = 0.5
-        return np.exp(-0.5 * np.sum(diffs ** 2, axis=2) / self.bandwidth ** 2)
-
-    def _statistic(self, X, Y):
-        N, _ = X.shape
-        M, _ = Y.shape
-        x_stat = np.sum(self._gaussian_covariance(X, X) - np.eye(N)) / (N * (N - 1))
-        y_stat = np.sum(self._gaussian_covariance(Y, Y) - np.eye(M)) / (M * (M - 1))
-        xy_stat = np.sum(self._gaussian_covariance(X, Y)) / (N * M)
-        return x_stat - 2 * xy_stat + y_stat
+        self.num_workers = num_workers
 
     def _embed(self, A1, A2):
+        if not is_symmetric(A1) or not is_symmetric(A2):
+            raise NotImplementedError()  # TODO asymmetric case
+
+        if self.n_components is None:
+            num_dims1 = select_dimension(A1)[0][-1]
+            num_dims2 = select_dimension(A2)[0][-1]
+            self.n_components = max(num_dims1, num_dims2)
+
         ase = AdjacencySpectralEmbed(n_components=self.n_components)
         X1_hat = ase.fit_transform(A1)
         X2_hat = ase.fit_transform(A2)
+
         if isinstance(X1_hat, tuple) and isinstance(X2_hat, tuple):
             X1_hat = np.concatenate(X1_hat, axis=-1)
             X2_hat = np.concatenate(X2_hat, axis=-1)
         elif isinstance(X1_hat, tuple) ^ isinstance(X2_hat, tuple):
             raise ValueError("Input graphs do not have same directedness")
+
         return X1_hat, X2_hat
 
-    def _median_heuristic(self, X1, X2):
+    def _medial_gaussian_kernel(self, x, workers=None):
```

Comment:
I'd say for consistency this function should have a signature 
 _medial_gaussian_kernel(self, X, Y=None, workers=None).
and behave similarly to metrics in scipy. not a big issue right now, since currently hyppo always sets parameters by position and only uses X.

Reply:
correction: metrics in sklearn

Reply:
agree. do you know anything about the `workers` part being necessary for a function passed into hyppo k-sample? I don't see anything about that in the hyppo docs

@sampan501

Reply:
yeah, I added these, actually. I chatted with Sambit earlier today, because my code kept breaking. basically, if one doesn't have them, there is an error of the following sort, at least with dcorr:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-127355934a25> in <module>
      1 ldt_dcorr = LatentDistributionTest("dcorr", metric="euclidean", n_bootstraps=1000)
----> 2 p_dcorr = ldt_dcorr.fit(A1, A2)
      3 ldt_mgc = LatentDistributionTest("mgc", metric="euclidean", n_bootstraps=1000)
      4 p_mgc = ldt_mgc.fit(A1, A2)
      5 print(p_dcorr, p_mgc)
~/Documents/Code/graspy/graspy/inference/latent_distribution_test.py in fit(self, A1, A2)
    236         y = np.array(X2_hat)
    237 
--> 238         data = self.test.test(x, y, reps=self.n_bootstraps, workers=self.num_workers, auto=False)
    239         self.sample_T_statistic_ = data[0]
    240         self.p_value_ = data[1]
~/.virtualenvs/graphs/lib/python3.7/site-packages/hyppo/ksample/ksamp.py in test(self, reps, workers, auto, *args)
    164 
    165         if self.indep_test_name in ["dcorr", "hsic"]:
--> 166             return self.indep_test.test(u, v, reps, workers, auto=auto)
    167         else:
    168             return self.indep_test.test(u, v, reps, workers)
~/.virtualenvs/graphs/lib/python3.7/site-packages/hyppo/independence/dcorr.py in test(self, x, y, reps, workers, auto, bias)
    210         else:
    211             if not self.is_distance:
--> 212                 x = self.compute_distance(x, workers=workers)
    213                 y = self.compute_distance(y, workers=workers)
    214                 self.is_distance = True
TypeError: dist_func() got an unexpected keyword argument 'workers'
```

Reply:
since he is "redoing the whole compute_distance thing anyway, it should be fixed in a future release", but as a temporary workaround we can just have functions that take a parameter "workers". whether they actually parallelize anything or not isn't as important (but might speed things up).

Reply:
but I should clarify that this doesn't mean that we shouldn't pass num_workers to hyppo, as this is not the only place where parallelization occurs.

Reply:
gotcha, whoops, i'll add back in